lib/pbio/drv/adc/adc_ev3.c

```diff
@@ -48,127 +102,94 @@ pbio_error_t pbdrv_adc_get_ch(uint8_t ch, uint16_t *value) {
+            shut_down_hack_done = 1;
+            break;
+        }
 
-    // Register the ISR in the Interrupt Vector Table.
-    IntRegister(SYS_INT_SPINT0, spi0_isr);
-    IntChannelSet(SYS_INT_SPINT0, 2);
-    IntSystemEnable(SYS_INT_SPINT0);
+        if (adc_soon) {
+            adc_soon = 0;
+            pbio_os_timer_set(&timer, ADC_SAMPLE_PERIOD);
+        } else {
+            // TODO: There should probably be a pbio OS function for this
+            timer.start += timer.duration;
+        }
 
-    // Reset.
-    SPIReset(SOC_SPI_0_REGS);
-    SPIOutOfReset(SOC_SPI_0_REGS);
+        // Do a sample of all channels
+        pbdrv_block_device_ev3_spi_begin_for_adc(
+            channel_cmd,
+            channel_data,
+            PBDRV_CONFIG_ADC_EV3_ADC_NUM_CHANNELS + PBDRV_CONFIG_ADC_EV3_NUM_DELAY_SAMPLES);
+        PBIO_OS_AWAIT_WHILE(state, pbdrv_block_device_ev3_is_busy());
 
-    // Mode.
-    uint32_t spipc0 = SPI_SPIPC0_SOMIFUN | SPI_SPIPC0_SIMOFUN | SPI_SPIPC0_CLKFUN | SPI_SPIPC0_SCS0FUN3;
-    SPIModeConfigure(SOC_SPI_0_REGS, SPI_MASTER_MODE);
-    SPIPinControl(SOC_SPI_0_REGS, 0, 0, (unsigned int *)&spipc0);
+        for (uint32_t i = 0; i < pbdrv_adc_callback_count; i++) {
+            pbdrv_adc_callbacks[i]();
+        }
+    }
 
-    // Config.
-    SPIClkConfigure(SOC_SPI_0_REGS, SOC_SYSCLK_2_FREQ, 2000000, SPI_DATA_FORMAT0);
-    SPIConfigClkFormat(SOC_SPI_0_REGS, SPI_CLK_OUTOFPHASE | 0x00000010, SPI_DATA_FORMAT0);
-    SPIDelayConfigure(SOC_SPI_0_REGS, 10, 10, 10, 10);
-    SPIIntLevelSet(SOC_SPI_0_REGS, SPI_RECV_INTLVL | SPI_TRANSMIT_INTLVL);
-    SPIDefaultCSSet(SOC_SPI_0_REGS, 8);
+    PBIO_OS_ASYNC_END(PBIO_SUCCESS);
+}
 
-    // Enable and loop around all channels.
-    SPIEnable(SOC_SPI_0_REGS);
+void pbdrv_adc_init(void) {
+    // Immediately go into async mode so that we can wait for the SPI flash driver.
+    // We *don't* want to block the initial init phase, or else things will deadlock.
 
-    process_start(&pbdrv_adc_process);
+    pbio_os_process_start(&pbdrv_adc_ev3_process, pbdrv_adc_ev3_process_thread, NULL);
 }
 
 void pbdrv_adc_update_soon(void) {
-    process_poll(&pbdrv_adc_process);
+    adc_soon = 1;
+    pbio_os_request_poll();
 }
 
-PROCESS_THREAD(pbdrv_adc_process, ev, data) {
-    PROCESS_EXITHANDLER(pbdrv_adc_exit());
-
-    static struct etimer etimer;
-
-    PROCESS_BEGIN();
-
-    etimer_set(&etimer, 10);
-    for (;;) {
-        PROCESS_WAIT_EVENT_UNTIL((ev == PROCESS_EVENT_TIMER && etimer_expired(&etimer)) || ev == PROCESS_EVENT_POLL);
-
-        channel_data_index = 0;
-        adc_busy = true;
-        SPIEnable(SOC_SPI_0_REGS);
-        SPIIntEnable(SOC_SPI_0_REGS, SPI_TRANSMIT_INT);
-        PROCESS_WAIT_EVENT_UNTIL(ev == PROCESS_EVENT_POLL && !adc_busy);
-
-        for (uint32_t i = 0; i < pbdrv_adc_callback_count; i++) {
-            pbdrv_adc_callbacks[i]();
-        }
-
-        etimer_reset(&etimer);
-    }
+void pbdrv_adc_ev3_configure_data_format() {
```

Comment:
When using DMA, we need to max out all of the delays with `SPIDelayConfigure()`. Even with this, we get slightly inaccurate readings because there isn't enough settle time between readings, so if you read a 0V channel followed by a 5V channel, the reading for the 5V channel is a bit less than 5V because it takes time for the voltage to transition from 0V to 5V when the mux is switched.

Slowing the the SCLK could probably help with this as well.

Reply:
The ADS7957 datasheet says that the sampling instant is controlled by the falling edge of the chip select, so it sounds like changing `WDELAY` (rather than the `SPIDELAY` delays) should fix this situation?

Am I correct in understanding that this is entirely an analog circuit problem and not a digital timing problem?

Reply:
Added maximum delays for all timing parameters